src/metabase/db/metadata_queries.clj

```diff
@@ -93,32 +93,47 @@
 
 (def TableRowsSampleOptions
   "Schema for `table-rows-sample` options"
-  (s/maybe {(s/optional-key :truncation-size) s/Int}))
+  (s/maybe {(s/optional-key :truncation-size) s/Int
+            (s/optional-key :rff)             s/Any}))
 
-(s/defn table-rows-sample :- (s/maybe si/TableSample)
-  "Run a basic MBQL query to fetch a sample of rows belonging to a Table. Can pass in optional `:truncation-size` for
-  text fields which will be honored only if the driver supports expressions."
+(defn- table-rows-sample-query
+  "Returns the mbql query to query a table for sample rows"
+  [table fields {:keys [truncation-size] :as _opts}]
+  (let [driver             (-> table table/database driver.u/database->driver)
+        text-fields        (filter (comp #{:type/Text} :base_type) fields)
+        field->expressions (when (and truncation-size (driver/supports? driver :expressions))
+                             (into {} (for [field text-fields]
+                                        [field [(str (gensym "substring"))
+                                                [:substring [:field-id (u/get-id field)] 1 truncation-size]]])))]
+    {:database   (:db_id table)
+     :type       :query
+     :query      {:source-table (u/get-id table)
+                  :expressions  (into {} (vals field->expressions))
+                  :fields       (vec (for [field fields]
+                                       (if-let [[expression-name _] (get field->expressions field)]
+                                         [:expression expression-name]
+                                         [:field-id (u/get-id field)])))
+                  :limit        max-sample-rows}
+     :middleware {:format-rows?           false
+                  :skip-results-metadata? true}}))
+
+(s/defn table-rows-sample
+  "Run a basic MBQL query to fetch a sample of rows of FIELDS belonging to a TABLE.
+
+  Options: a map of
+  `:truncation-size`: [optional] size to truncate text fields if the driver supports expressions.
+  `:rff`: [optional] a reducing function to reduce over the result set in the the query-processor rather than
+  realizing the whole collection"
```

Comment:
Do we actually need both options? Currently `table-rows-sample` is only used in one place. If we ever need a realised version it might be better to have an adapter function that does the transduction.

Reply:
that's true. and its kind of a footgun to start doing table samples without it as you end up with memory pressure.

That said, the testing strategy is _WAY_ nicer by just mocking out what the table-rows-sample function returns versus hijacking the query processor to just use `rff` over the return values.